lib/plugins/processor/post.ts

```diff
@@ -268,29 +269,41 @@ function processAsset(ctx: Hexo, file: _File) {
   const PostAsset = ctx.model('PostAsset');
   const Post = ctx.model('Post');
   const id = file.source.substring(ctx.base_dir.length);
-  const doc = PostAsset.findById(id);
+  const postAsset = PostAsset.findById(id);
 
-  if (file.type === 'delete') {
-    if (doc) {
-      return doc.remove();
+  if (file.type === 'delete' || Post.length === 0) {
+    if (postAsset) {
+      return postAsset.remove();
     }
-
     return;
   }
 
-  // TODO: Better post searching
-  const post = Post.toArray().find(post => file.source.startsWith(post.asset_dir));
-  if (post != null && (post.published || ctx._showDrafts())) {
+  const savePostAsset = (post: PostSchema) => {
     return PostAsset.save({
       _id: id,
       slug: file.source.substring(post.asset_dir.length),
       post: post._id,
       modified: file.type !== 'skip',
       renderable: file.params.renderable
     });
+  };
+
+  if (postAsset) {
+    // `postAsset.post` is `Post.id`.
+    const post = Post.findById(postAsset.post);
+    if (post != null && (post.published || ctx._showDrafts())) {
+      return savePostAsset(post);
+    }
+  }
+
+  const assetDir = id.slice(0, id.lastIndexOf(sep));
+  const post = Post.findOne(p => p.asset_dir.endsWith(posix.join(assetDir, '/')));
+  if (post != null && (post.published || ctx._showDrafts())) {
+    return savePostAsset(post);
   }
 
-  if (doc) {
-    return doc.remove();
+  // NOTE: Probably, unreachable.
```

Comment:
Probably, will not be reached here unless the user intentionally changes the values in `db.json`. But, I'm not sure about this, so I haven't removed it.